src/models/shelf/function.ts

```diff
@@ -0,0 +1,158 @@
+import {isShortWildcard, isWildcard, Wildcard} from 'compassql/build/src/wildcard';
+import {AGGREGATE_OPS, AggregateOp} from 'vega-lite/build/src/aggregate';
+import {TimeUnit, TIMEUNITS} from 'vega-lite/build/src/timeunit';
+import {contains} from 'vega-lite/build/src/util';
+import {toSet} from 'vega-util';
+
+export type ShelfFunction = AggregateOp | 'bin' | TimeUnit | undefined;
+
+const AGGREGATE_INDEX = toSet(AGGREGATE_OPS);
+const TIMEUNIT_INDEX = toSet(TIMEUNITS);
+
+function isAggregate(fn: string): fn is AggregateOp {
+  return AGGREGATE_INDEX[fn];
+}
+
+function isTimeUnit(fn: string): fn is TimeUnit {
+  return TIMEUNIT_INDEX[fn];
+}
+
+export function isShelfFunction(fn: string): fn is ShelfFunction {
+  return fn === 'bin' ||
+    fn === undefined || fn === null || // check null for duplicate
+    isAggregate(fn) || isTimeUnit(fn);
+}
+
+export function toFunctionMixins(fn: ShelfFunction | Wildcard<ShelfFunction>):
+  Pick<FieldQuery, 'aggregate' | 'timeUnit' | 'bin' | 'hasFn'> {
+
+  if (isWildcard(fn)) {
+    const aggregates: AggregateOp[] = [];
+    const timeUnits: TimeUnit[] = [];
+    let hasBin: boolean = true;
+    let hasNoFn: boolean = true;
+
+    for (const f of fn.enum) {
+      if (isAggregate(f)) {
+        aggregates.push(f);
+      } else if (isTimeUnit(f)) {
+        timeUnits.push(f);
+      } else if (f === 'bin') {
+        hasBin = true;
+      } else if (f === undefined || f === null) {
+        // Check for null just in case things get copied
+        hasNoFn = true;
+      } else {
+        throw new Error('Invalid function ' + f);
+      }
+    }
+
+    const functionTypeCount = (aggregates.length > 0 ? 1 : 0) +
+      (timeUnits.length > 0 ? 1 : 0) +
+      (hasBin ? 1 : 0);
+
+    const enumerateUndefined = functionTypeCount > 1 || hasNoFn;
+    const baseEnum: Array<undefined> = enumerateUndefined ? [undefined] : [];
+    const hasFn = !hasNoFn;
+
+
+    return {
+      ...(aggregates.length > 0 ? {
+        aggregate: {enum: [].concat(baseEnum, aggregates)}
+      } : {}),
+      ...(timeUnits.length > 0 ? {
+        timeUnit: {enum: [].concat(baseEnum, timeUnits)}
+      } : {}),
+      ...(hasBin ? {
+        bin: {
+          enum: (enumerateUndefined ? [false] : []).concat([true])
+          // TODO: deal with bin params
+        }
+      } : {}),
+      ...(hasFn ? {hasFn} : {})
+    };
+  } else if (isAggregate(fn)) {
+    return {aggregate: fn};
+  } else if (fn === 'bin') {
+    return {bin: true};
+  } else if (isTimeUnit(fn)) {
+    return {timeUnit: fn};
+  }
+  return {};
+}
+
+function excludeUndefined(fn: string) {
+  if (!isShelfFunction) {
+    console.warn(`Invalid function ${fn} dropped`);
+    return false;
+  }
+  return fn !== undefined && fn !== null;
+}
+
+export function fromFieldQueryFunctionMixins(
+  fieldQParts: Pick<FieldQuery, 'aggregate' | 'timeUnit' | 'bin' | 'hasFn'>
+): ShelfFunction | Wildcard<ShelfFunction> {
+
+  // FIXME make this a parameter
```

Comment:
Please file an issue that we should make this a parameter.  (Let's do it later.)